samples/nrf9160/modem_trace_flash/src/trace_flash.c

```diff
@@ -0,0 +1,181 @@
+#define TRACE_SIZE FLASH_AREA_SIZE(modem_trace)
+#define BUF_LEN 1024
+
+static const struct flash_area *modem_trace_area;
+static const struct device *flash_dev;
+static struct stream_flash_ctx stream;
+static uint8_t stream_buf[BUF_LEN];
+
+// Persist trace size in .noinit RAM to survive warm boot.
+static __noinit uint32_t magic;
+static __noinit uint32_t traces_size;
+
+static int64_t tot_write_time;
+static trace_backend_processed_cb trace_processed_callback;
+
+static int internal_flash_write(const void *data, size_t len, bool flush)
+{
+	int64_t uptime_ref = k_uptime_get();
+	int64_t write_time;
+
+	int err = stream_flash_buffered_write(&stream, data, len, flush);
+
+	if (err != 0) {
+		LOG_ERR("stream_flash_buffered_write error %d", err);
+		return err;
+	}
+
+	write_time = k_uptime_delta(&uptime_ref);
+	tot_write_time += write_time;
+
+	traces_size += len;
+
+	return (int)len;
+}
+
+static const uint32_t magic_val = 0x152ac523;
+int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
+{
+	int err;
+
+	LOG_DBG("trace_backend_init\n");
+	if (trace_processed_cb == NULL) {
+		return -EFAULT;
+	}
+
+	trace_processed_callback = trace_processed_cb;
+
+	if (magic != magic_val) {
+		LOG_DBG("Initializing .noinit\n");
+		magic = magic_val;
+		traces_size = 0;
+	}
+
+	traces_size = 0;
+
+	err = flash_area_open(FLASH_AREA_ID(modem_trace), &modem_trace_area);
+	if (err) {
+		LOG_ERR("flash_area_open error:  %d", err);
+		return -ENODEV;
+	}
+
+	err = flash_area_has_driver(modem_trace_area);
+	if (err != 1) {
+		LOG_ERR("flash_area_has_driver error: %d\n", err);
+	}
+
+	flash_dev = flash_area_get_device(modem_trace_area);
+	if (flash_dev == NULL) {
+		LOG_ERR("Failed to get flash device\n");
+		return -ENODEV;
+	}
+
+	err = stream_flash_init(&stream, flash_dev, stream_buf, BUF_LEN, TRACE_OFFSET, TRACE_SIZE,
+		NULL);
+	if (err) {
+		LOG_ERR("stream_flash_init error: %d", err);
+		return err;
+	}
+
+	// erase flash partition first to write without interruptions while tracing
+	int64_t erase_before = k_uptime_get();
+
+	err = flash_erase(flash_dev, TRACE_OFFSET, TRACE_SIZE);
+	if (err) {
+		LOG_ERR("flash_erase error: %d", err);
+		return err;
+	}
+	int64_t erase_time = k_uptime_delta(&erase_before);
+
+	LOG_DBG("Erased %lu bytes in %lld ms", (unsigned long) TRACE_SIZE, erase_time);
+	LOG_DBG("Erase speed %lld kb/s ", (TRACE_SIZE * 8 / erase_time));
+
+	tot_write_time = 0;
+
+	LOG_DBG("Modem trace flash backend initialized\n");
+
+	return 0;
+}
+
+int trace_backend_deinit(void)
+{
+	LOG_DBG("Flash trace backend deinitialized\n");
+
+	return 0;
+}
+
+static bool has_flushed;
+
+int trace_backend_write(const void *data, size_t len)
+{
+	if (has_flushed) {
+		trace_processed_callback(len);
+		return -EIO;
+	}
+
+	int ret = internal_flash_write(data, len, false);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	trace_processed_callback(len);
+	if (ret < 0) {
+		LOG_ERR("Unexpected long wait for flush mutex");
+		return ret;
+	}
```

Comment:
```suggestion
	ret = trace_processed_callback(len);
	if (err< 0) {
		LOG_ERR("trace_processed_callback returns %d", ret);
		return ret;
	}
```

Reply:
Thanks, the code here was not correct, but we do need to return an error when the callback returns an error. That means we called the callback with an invalid length, so we should return an error and abort the modem tracing (which nrf_modem_lib_trace handles).